tuf/fetcher.py

```diff
@@ -0,0 +1,188 @@
+  Teodora Sechkova <tsechkova@vmware.com>
+
+<Started>
+  December 14, 2020
+
+<Copyright>
+  See LICENSE-MIT OR LICENSE for licensing information.
+
+<Purpose>
+  Provides an interface for network IO abstraction.
+"""
+
+import abc
+import requests
+import six
+import logging
+import time
+
+import urllib3.exceptions
+import tuf.exceptions
+import tuf.settings
+
+logger = logging.getLogger(__name__)
+
+
+class FetcherInterface():
+  """
+  <Purpose>
+  Defines an interface for abstract network download which can be implemented
+  for a variety of network libraries and configurations.
+  """
+  __metaclass__ = abc.ABCMeta
+
+
+  @abc.abstractmethod
+  def fetch(self, url, required_length):
+    """
+    <Purpose>
+      Fetches the contents of HTTP/HTTPS url from a remote server up to
+      required_length and returns a bytes iterator.
+
+    <Arguments>
+      url:
+        A URL string that represents the location of the file.
+
+      required_length:
+        An integer value representing the length of the file in bytes.
+
+    <Exceptions>
+      tuf.exceptions.SlowRetrievalError, if a timeout occurs while receiving
+      data from a server
+
+    <Returns>
+      A bytes iterator
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+
+class RequestsFetcher(FetcherInterface):
+  """
+  <Purpose>
+    A concrete implementation of FetcherInterface based on the Requests
+    library.
+  """
+
+  def __init__(self):
+    # From http://docs.python-requests.org/en/master/user/advanced/#session-objects:
+    #
+    # "The Session object allows you to persist certain parameters across requests.
+    # It also persists cookies across all requests made from the Session instance,
+    # and will use urllib3's connection pooling. So if you're making several
+    # requests to the same host, the underlying TCP connection will be reused,
+    # which can result in a significant performance increase (see HTTP persistent
+    # connection)."
+    #
+    # NOTE: We use a separate requests.Session per scheme+hostname combination, in
+    # order to reuse connections to the same hostname to improve efficiency, but
+    # avoiding sharing state between different hosts-scheme combinations to
+    # minimize subtle security issues. Some cookies may not be HTTP-safe.
+    self._sessions = {}
+
+
+  def fetch(self, url, required_length):
+    # Get a customized session for each new schema+hostname combination.
+    session = self._get_session(url)
+
+    # Get the requests.Response object for this URL.
+    #
+    # Defer downloading the response body with stream=True.
+    # Always set the timeout. This timeout value is interpreted by requests as:
+    #  - connect timeout (max delay before first byte is received)
+    #  - read (gap) timeout (max delay between bytes received)
+    response = session.get(url, stream=True,
+        timeout=tuf.settings.SOCKET_TIMEOUT)
+    # Check response status.
+    response.raise_for_status()
+
+    def chunks():
```

Comment:
could add a comment describing the trick here: chunks is the actual iterator that gets returned from fetch() after response object has been initialized (and most importantly after connect() has happened)